dlt/destinations/impl/filesystem/sql_client.py

```diff
@@ -0,0 +1,250 @@
+                if aws_creds.endpoint_url
+                else "s3.amazonaws.com"
+            )
+            self._conn.sql(f"""
+            CREATE OR REPLACE {persistent_stmt} SECRET {secret_name} (
+                TYPE S3,
+                KEY_ID '{aws_creds.aws_access_key_id}',
+                SECRET '{aws_creds.aws_secret_access_key}',
+                REGION '{aws_creds.region_name}',
+                ENDPOINT '{endpoint}',
+                SCOPE '{scope}'
+            );""")
+
+        # azure with storage account creds
+        elif self.fs_client.config.protocol in ["az", "abfss"] and isinstance(
+            self.fs_client.config.credentials, AzureCredentialsWithoutDefaults
+        ):
+            azsa_creds = self.fs_client.config.credentials
+            self._conn.sql(f"""
+            CREATE OR REPLACE {persistent_stmt} SECRET {secret_name} (
+                TYPE AZURE,
+                CONNECTION_STRING 'AccountName={azsa_creds.azure_storage_account_name};AccountKey={azsa_creds.azure_storage_account_key}',
+                SCOPE '{scope}'
+            );""")
+
+        # azure with service principal creds
+        elif self.fs_client.config.protocol in ["az", "abfss"] and isinstance(
+            self.fs_client.config.credentials, AzureServicePrincipalCredentialsWithoutDefaults
+        ):
+            azsp_creds = self.fs_client.config.credentials
+            self._conn.sql(f"""
+            CREATE OR REPLACE {persistent_stmt} SECRET {secret_name} (
+                TYPE AZURE,
+                PROVIDER SERVICE_PRINCIPAL,
+                TENANT_ID '{azsp_creds.azure_tenant_id}',
+                CLIENT_ID '{azsp_creds.azure_client_id}',
+                CLIENT_SECRET '{azsp_creds.azure_client_secret}',
+                ACCOUNT_NAME '{azsp_creds.azure_storage_account_name}',
+                SCOPE '{scope}'
+            );""")
+        elif persistent:
+            raise Exception(
+                "Cannot create persistent secret for filesystem protocol"
+                f" {self.fs_client.config.protocol}. If you are trying to use persistent secrets"
+                " with gs/gcs, please use the s3 compatibility layer."
+            )
+
+        # native google storage implementation is not supported..
+        elif self.fs_client.config.protocol in ["gs", "gcs"]:
+            logger.warn(
+                "For gs/gcs access via duckdb please use the gs/gcs s3 compatibility layer. Falling"
+                " back to fsspec."
+            )
+            self._conn.register_filesystem(self.fs_client.fs_client)
+
+        # for memory we also need to register filesystem
+        elif self.fs_client.config.protocol == "memory":
+            self._conn.register_filesystem(self.fs_client.fs_client)
+
+    def open_connection(self) -> duckdb.DuckDBPyConnection:
+        # we keep the in memory instance around, so if this prop is set, return it
+        if not self._conn:
```

Comment:
I've changed it a bit to keep the in memory connection in this class and it can be treated as an external database in the credentials, this way the borrowing will work just the same way it always does and the instance should go away when this class gets garbarge collected.